### PR TITLE
azure - support multiple azure clouds

### LINKED ISSUE #5166: azure - How can config it support for ‘AzureChinaCloud’ environment?
### How can config it support for ‘AzureChinaCloud’ environment? 

I have an Azure China subscription account. I want to support it in China. How do I configure it?

### PR DESCRIPTION
**Pull request by wsf11**
Support multiple Azure Cloud types by overloading the `--region` flag

The main change is adding `base_url` as a parameter when grabbing the Client, and making sure the correct URL was given to the authentication mechanisms. 

Got rid of constant endpoints to dynamically get correct endpoints/Cloud region. 

Known Issues to be addressed after: 
1. Using MSI & SP with separate Clouds requires passing in the Cloud Endpoint. CLI works for other Clouds which is a good starting point for testing
2. KeyVault Base URL will have to accessed in a different way as the URL will not be able to be dynamic for Clouds as is. 

closes #5166

### CODE REVIEW

File: tools/c7n_azure/c7n_azure/provider.py
Comment (kapilt) [inline tools/c7n_azure/c7n_azure/provider.py]: probably should go in constants with the rest? given its using an import, I think its fine either way tbh

File: tools/c7n_azure/tests_azure/test_provider.py
Comment (kapilt) [inline tools/c7n_azure/tests_azure/test_provider.py]: are these values well known for azure cli? just wondering if we should take an opportunity to shorten, or if the values are well known (ie. strip the common cloud suffix, possibly other transforms).

Comment (wsf11) [reply tools/c7n_azure/tests_azure/test_provider.py]: Running `az cloud list` returns the names: AzureCloud, AzureChinaCloud, AzureUSGovernment, and AzureGermanCloud as the names. I believe they are well known. Probably could change `USGov` -> `USGovernment` for consistency

**Comment by kapilt** [comment]
one other thing that would be good to see here per comment in original issue, is an explicit early fail if a policy file has multiple providers.

File: tools/c7n_azure/tests_azure/test_provider.py
Comment (kapilt) [reply tools/c7n_azure/tests_azure/test_provider.py]: we should probably default to the well known vocabulary,  we're printing the value valid values on a bad one, so discoverability is good without resorting to a secondary source.

File: tools/c7n_azure/c7n_azure/session.py
Comment (stefangordon) [inline tools/c7n_azure/c7n_azure/session.py]: I'm not sure these two things are different?

File: tools/c7n_azure/c7n_azure/provider.py
Comment (wsf11) [reply tools/c7n_azure/c7n_azure/provider.py]: We can log an error. I think it may be easy to be confused because with AWS you can pass in multiple regions at once and we are overloading the term to be used per Cloud

File: tools/c7n_azure/c7n_azure/query.py
Comment (wsf11) [reply tools/c7n_azure/c7n_azure/query.py]: So it seems that the only time this will be hit is for `KeyVault` child resources because we set the resource type to be `https://vault.azure.net`. ie[ here](https://github.com/cloud-custodian/cloud-custodian/blob/master/tools/c7n_azure/c7n_azure/resources/key_vault_keys.py#L93)

This also means as is, actions on KeyVault child resources won't work in other Clouds because we will be requesting against `https://vault.azure.net` rather than say `https://vault.azure.cn`

Comment (wsf11) [reply tools/c7n_azure/c7n_azure/query.py]: I initially changed this because I assumed that the resource type would change to say `china.management` so the constant `management` would catch it, but that doesn't seem to be the case

Comment (wsf11) [reply tools/c7n_azure/c7n_azure/query.py]: That could be in the scope of a separate PR as this PR is to initially unblock other Clouds

Comment (stefangordon) [reply tools/c7n_azure/c7n_azure/query.py]: I think we should be using this https://github.com/Azure/msrestazure-for-python/blob/master/msrestazure/azure_cloud.py#L131 and remove all references to specific endpoint constants from our constants file. 

Session can use that table to lookup all the endpoints itself I think, and instead of passing it endpoints in the constructor maybe we can just pass the keyword name ("management") or whatever.

Comment (wsf11) [reply tools/c7n_azure/c7n_azure/query.py]: We can do that. We will still need to pass in a set of endpoints or Cloud name so Session has knowledge which endpoints to grab from

**Comment by wsf11** [comment]
> one other thing that would be good to see here per comment in original issue, is an explicit early fail if a policy file has multiple providers.

I can work on it. Should that be in a separate PR?

File: tools/c7n_azure/c7n_azure/provider.py
Comment (kapilt) [inline tools/c7n_azure/c7n_azure/provider.py]: can we rename this to endpoint_url or cloud_endpoint_url  .. cloud feels a bit nebulous

please also explicitly set options['region'] to the so it can be used as a cache key per #5447

**Comment by kapilt** [comment]
> > one other thing that would be good to see here per comment in original issue, is an explicit early fail if a policy file has multiple providers.
> 
> I can work on it. Should that be in a separate PR?

a separate pr for that sounds good.

File: tools/c7n_azure/c7n_azure/session.py
Comment (logachev) [inline tools/c7n_azure/c7n_azure/session.py]: nit: `cloud_endpoints` is enough because `AZURE_PUBLIC_CLOUD` is a default value for the parameter

File: tools/c7n_azure/c7n_azure/session.py
Comment (logachev) [inline tools/c7n_azure/c7n_azure/session.py]: Where do we set `base_url`?

File: tools/c7n_azure/c7n_azure/storage_utils.py
Comment (logachev) [inline tools/c7n_azure/c7n_azure/storage_utils.py]: Should we add `StorageEndpoint` property to the `Session` to hide the internals (like `endpoints.suffixes`)

File: tools/c7n_azure/tests_azure/test_azure_utils.py
Comment (logachev) [inline tools/c7n_azure/tests_azure/test_azure_utils.py]: I think it is better to construct an expected endpoint manually. Since `get_keyvault_secret` uses same function, if you accidentally edit the function test will still be successful.

File: tools/c7n_azure/tests_azure/test_azure_utils.py
Comment (logachev) [inline tools/c7n_azure/tests_azure/test_azure_utils.py]: nit: Can you add some other cloud too?

File: tools/c7n_azure/tests_azure/test_provider.py
Comment (logachev) [inline tools/c7n_azure/tests_azure/test_provider.py]: You can use `@pytest.mark.parametrize` to pass a set of parameters and have single test

File: tools/c7n_azure/c7n_azure/constants.py
Comment (logachev) [inline tools/c7n_azure/c7n_azure/constants.py]: I'd probably consider different names for these.. It is not an endpoint, maybe some resource type? @stefangordon what do you think?

In the code we assign its value to the `resource` property.. So feels confusing.

File: tools/c7n_azure/c7n_azure/session.py
Comment (wsf11) [reply tools/c7n_azure/c7n_azure/session.py]: When `get_session_factory` is called [here](https://github.com/cloud-custodian/cloud-custodian/blob/master/tools/c7n_azure/c7n_azure/provider.py#L43), it initially sets `cloud_endpoints=None`. So when the `resolve_auth_endpoint` is called in the constructor, it is called with empty cloud endpoints causing an error to be thrown.

File: tools/c7n_azure/c7n_azure/session.py
Comment (wsf11) [reply tools/c7n_azure/c7n_azure/session.py]: Good catch, this was left over from initial draft

File: tools/c7n_azure/c7n_azure/constants.py
Comment (wsf11) [reply tools/c7n_azure/c7n_azure/constants.py]: The endpoints are used to get a token for each resource, so it felt like an auth endpoint. We assign it `auth_endpoint` in the Session constructor, but we assign the `vault` to the `resource` property which I agree is confusing. They are all used to dynamically get an auth endpoint depending on the Cloud. 

`STORAGE_AUTH_ENDPOINT` will stay the same regardless on the cloud. For storage, you pass in the correct suffix. As seen in `storage_utils.py`

File: tools/c7n_azure/c7n_azure/provider.py
Comment (stefangordon) [inline tools/c7n_azure/c7n_azure/provider.py:83]: If you want to avoid keeping two lists of cloud names, `", ".join(region_to_cloud.keys())`

File: tools/c7n_azure/c7n_azure/session.py
Comment (stefangordon) [reply tools/c7n_azure/c7n_azure/session.py:71]: I'd try to look more into where/why this is happening, is the `partial` causing a problem?  It feels weird to have the constructor defensively try to solve problems with callers sending invalid parameters.   It looked like it was set correctly everywhere (or not passed) but I haven't run it :)

File: tools/c7n_azure/c7n_azure/session.py
Comment (stefangordon) [inline tools/c7n_azure/c7n_azure/session.py:75]: I find this a bit hard to follow, a combination of the way it is resolved/used, the overrides, and the inconsistent naming.  Calling into the resolve method to set the class level attributes isn't ideal as its harder to read and test.  One proposal that I think simplifies everything a bunch:

Get rid of `keyvault_auth_override`, and just keep both the keyword and the resolved value in two fields here (rename constructor `auth_endpoint` to `resource_endpoint_type`).  Return the single value from `resolve_auth_endpoint` (and optionally rename to `get_auth_endpoint`.

```
self.cloud_endpoints = cloud_endpoints
self.resource_endpoint_type = resource_endpoint_type
self.resource_endpoint = self.get_auth_endpoint(resource_endpoint_type)
self.storage_endpoint = self.cloud_endpoints.suffixes.storage_endpoint
```

Then for the keyvault check you can just keep doing basically the old way that is cloud agnostic.
```
if self.resource_endpoint_type == constants.keyvault_something:
```

File: tools/c7n_azure/c7n_azure/session.py
Comment (stefangordon) [inline tools/c7n_azure/c7n_azure/session.py:326]: the `self.endpoints.endpoints` stuff isn't ideal - per recommendation above I'd rename to something else, even `cloud_endpoints` so it makes a little more sense - I know it comes at expense of restrictive line length linting making life hard...

**Comment by rachgupt** [comment]
Hi,
Were you able to get it working for Gov and China? We have similar issue in https://github.com/cloud-custodian/cloud-custodian/issues/6072

**Comment by cpollard0** [comment]
@stefangordon , @kapilt - Is this worth reviving or if I wanted to add in Azure Gov support I should start a new branch?

**Comment by stefangordon** [comment]
@wsf11 If you get a chance today please sign the new CLA (comment above) as I'm going to try to merge this :)

**Comment by stefangordon** [comment]
Ok I believe this is good to go now.  I've done some manual testing with Azure CLI auth, SP auth and UAI Functions as well.

**Comment by wsf11** [comment]
> @wsf11 If you get a chance today please sign the new CLA (comment above) as I'm going to try to finish/merge this 👍

Done!

**Review by stefangordon — APPROVED** [review]
I implemented my own code review feedback on this one, so will let someone else glance over that last commit and approve before merging.

**Review by logachev — APPROVED** [review]

**Pull request merged**
